src/query/algorithms/entityresolution/EntityResolver.cpp

```diff
@@ -0,0 +1,502 @@
+ */
+vector<map<int, string>>
+EntityResolver::createFilters(map<int, vector<string>> entityData, map<int, vector<int>> neighborhoodData,
+                              int filterSize = 256,
+                              int numHashes = 4) {
+    //Create bloom filters
+    cout << "Creating filters" << endl;
+    Utils utils;
+    map<int, string> attrFilters;
+    map<int, string> structFilters;
+    //For each entity create attr and structural bloom filters
+    for (const auto &entity: entityData) {
+    //Create attr bloom filter
+        BloomFilter attrFilter(filterSize, numHashes);
+        vector<string> attributes = entity.second;
+    //Add node attributes to bloom filter
+        for (auto attr: attributes) {
+            cout << attr << endl;
+            attrFilter.insert(attr);
+        }
+    //Convert bloom filter to appropriate string
+        string filterStr = attrFilter.m_bits.to_string();
+        cout << filterStr << endl;
+        filterStr = utils.replace(filterStr, "0", ",0");
+        filterStr = utils.replace(filterStr, "1", ",1");
+        attrFilters[entity.first] = filterStr;
+
+    //Create structural filter
+        BloomFilter structFilter(256, 4);
+    //For each neighbour add selected attribute to bloom filter
+        for (auto neighbour: neighborhoodData[entity.first]) {
+            string selectedAttr = entityData[neighbour][0];
+            structFilter.insert(selectedAttr);
+        }
+    //Convert bloom filter to appropriate string
+        filterStr = structFilter.m_bits.to_string();
+        filterStr = utils.replace(filterStr, "0", ",0");
+        filterStr = utils.replace(filterStr, "1", ",1");
+        structFilters[entity.first] = filterStr;
+    }
+
+    return {attrFilters, structFilters};
+}
+
+/**
+ * Cluster created and collected Bloom filters of a given graph ID into a desired number of clusters
+ * @param graphID Graph ID of corresponding graph
+ * @param partitionCount The count of the (collected) partitions for the graph
+ * @param noClusters Desired clustering amount
+ */
+void EntityResolver::clusterFilters(string graphID, int partitionCount, int noClusters) {
+
+    Utils utils;
+    int noWorkers = stoi(utils.getJasmineGraphProperty("org.jasminegraph.server.nworkers"));
+    string partitionID = "0"; // use first worker related partitons
+    std::string attrFilterPath = "";
+    //std::string structFilterPath = "";
+
+    attrFilterPath =
+            utils.getJasmineGraphProperty("org.jasminegraph.server.instance.entityresolutionfolder") + "AttrFilters_" +
+            graphID + "_" + partitionID + ".txt";
+    //structFilterPath = utils.getJasmineGraphProperty("org.jasminegraph.entity.location")+"structfilters_" + graphID + "_" + partitionID + ".txt";
+
+    Mat<float> data;
+    data.load(attrFilterPath);
+
+    for (int i = 1; i < partitionCount; i++) {
+
+        string filepath = "attrfilters_" + graphID + "_" + to_string(i) + ".txt";
```

Comment:
Where do we have these attribute files? I hope those are loaded from JasmineGraph local store.

Reply:
Filepath changed

Reply:
Ok